src/Components/GEOSldas_GridComp/GEOSlandassim_GridComp/clsm_ensupd_upd_routines.F90

```diff
@@ -4000,6 +4053,225 @@ subroutine cat_enkf_increments(                               &
+                      
+                      cat_progn_incr(n, n_e)%wesn(i) = &
+                           ( Observations(ind_obs(1))%obs - &
+                           asnow_ensavg(n)/model_threshold ) * (wesn_added/float(N_snow))
+                      
+                      !if(logit) write(logunit,*) 'cat_progn_incr(n_n_e)%wesn = ', cat_progn_incr(n, n_e)%wesn(i)
+                      
+                      ! b) snow depth 
+
+                      !if(logit) write(logunit,*) 'cat_progn(n,n_e)%sndz(i)', cat_progn(n,n_e)%sndz(i)
+
+                      if(cat_progn(n,n_e)%sndz(i) > eps) then
+                         
+                         ! Determine the fractional snow coverage
+                         wesn_sum_old(n,n_e) = sum(cat_progn(n,n_e)%wesn)
+                         areasc_old(  n,n_e) = min(wesn_sum_old(n,n_e)/wemin,1.)
+                         
+                         ! if(logit) write(logunit,*) 'wesn_sum_old(n,n_e)', wesn_sum_old(n,n_e)
+                         ! if(logit) write(logunit,*) 'areasc_old(n,n_e)', areasc_old(n,n_e)
+                         
+                         ! Estimate the density of the ld layers of snow
+                         dens_layer = max(cat_progn(n,n_e)%wesn(i)/ &
+                              (cat_progn(n,n_e)%sndz(i)*areasc_old(n,n_e)),RHOFS)
+                         
+                         !if(logit) write(logunit,*) 'dens_layer', dens_layer
+                         
+                         cat_progn_incr(n,n_e)%sndz(i)=&
+                              cat_progn_incr(n,n_e)%wesn(i)/dens_layer
+                         
+                         !if(logit) write(logunit,*) 'cat_progn_incr(n,n_e)%sndz:', cat_progn_incr(n,n_e)%sndz(i)
+                         
+                      else
+                         dens_layer = RHOFS
+                         cat_progn_incr(n,n_e)%sndz(i) = cat_progn_incr(n,n_e)%wesn(i)/dens_layer
+                      end if
+
+                      ! c) snow heat content
+                      
+                      !if(logit) write(logunit,*) 'Deduction:', deduction(n)
+                      !original plan : use the min function --> continuously giving the floating error
+                      !TSN = min( deduction(n), 0.0)
+                      !Second options : replace the min function with if-else statement 
+                      if (deduction(n) .gt. 0.0) then
+                         TSN = 0.0
+                      else
+                         TSN = deduction(n)
+                      end if
+                      !if(logit) write(logunit,*) 'TSN = ',TSN
+                      !if(logit) write(logunit,*) 'wesn increment', cat_progn_incr(n,n_e)%wesn(i)
+                      !if(logit) write(logunit,*) 'cpw_pub: ', cpw_pub, 'MAPL_ALHF', MAPL_ALHF
+                      cat_progn_incr(n,n_e)%htsn(i) = real((TSN*cpw_pub-MAPL_ALHF))*cat_progn_incr(n,n_e)%wesn(i)
```

Comment:
This looks to be terribly wrong, I'm afraid, and is a carry-over from Ally's implementation.  The idea of the snow analysis is to compute and apply increments for SWE first.  Then we need to adjust snow depth and snow heat content such that the snow density and snow temperature remain constant, resp.  If there wasn't snow prior to the analysis, we probably want to use what is appropriate for freshly fallen snow.  
If I'm reading the existing code correctly, the snow heat content is **always** based on the difference between the **air** temperature and the freezing point.  This is **not** what it should be.  
If there was some snow prior to the analysis, the snow heat content is not updated correctly.  To do this correctly, we need to compute the temperature of the snow before the analysis, and then find out the heat content so that the updated snow has the same temperature before and after the analysis.  This is not trivial because the ice fraction probably also plays a role.  
It's possible but unlikely that this causes the floating point that @jpark50 experienced.